### PR TITLE
Work around unimplemented accNavigate when navigating to previous and next objects

### LINKED ISSUE #7178: Provide workaround for IAccessible implementations with broken next/previous object retrieval
### Steps to reproduce:
1. Go to the Windows 10 settings app
2. Type this in the python console:
    > nav.appModule.isBadUIAWindow=lambda hwnd: True
3. Focus the categories list
4. Try to use next/previous object navigation commands

### Expected behavior:
You should be able to navigate the list using object nav

### Actual behavior:
NVDA reports no next/previous element. However, if you do nav.parent.children in the python console, you'll realise that there are multiple list items

### Remarks
I do realise the str sounds a bit strange, why should one ever want to forcefully disable UIA altogether in the settings app? However, I've seen another app used in a corporate environment (IAccessible only), in which I saw very similar behaviour.
### System configuration:
NVDA version:
next-14047,eae024d0

NVDA Installed

Windows version:
Windows 10 Version 1703 (OS Build 15063.296)

### Other questions:

Does the issue still occur after restarting your PC?
Yes

Have you tried any other versions of NVDA?
No

CC @MichaelDCurran @josephsl

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
fixes #7178

### Background
[As stated by Microsoft](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318473%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396):

> Note  The accNavigate method is deprecated and should not be used. Clients should use other methods and properties such as AccessibleChildren, get_accChild, get_accParent, and IEnumVARIANT.

In the current implementation, NVDA relies solely on accNavigate to retrieve the previous and next object when using object navigation. However, as Microsoft stated its deprecated state, we must use retrieve the previous and next objects by other means if accNavigate doesn't yield an object even though there is a previous/next object in the tree. Steps to reproduce a situation where this is the case are provided in #7178.

### Description of how this pull request fixes the issue:
Some discussion has taken place with @jcsteh about this. After some consideration, I decided to keep the current implementation as conservative as possible.

1. the firstChild and lastChild properties also use accNavigate as a first approach to retrieve the first/last object. When accNavigate fails for these, this is a sign that accNavigate might also return nothing for previous/next object retrival on the first or last child.
2. If first/lastChild yield an object by other means than accNavigate, the returned object has its parent and the index inside its parent cached on the object if it is MSAA. For IAccessible2, this workaround is disabled for now.
3. Next/previous use the cached parent and the indexInParent properties to retrieve a sibling object if accNavigate fails. This means that the workaround for next/previous is only activated when indexInParent is present, and for MSAA objects, this is only the case when set in firstChild/lastChild earlier. Thus, obj.firstChild.next might yield an object, while obj.children[0].next will not. That's the conservative part. I only intended to fix issues regarding direct object navigation. See known issues below for further deliberations.

### Testing performed:
Followed the #7178 steps to reproduce, which allowed me to navigate to objects I couldn't reach earlier.

### Known issues and limitations
1. As stated, obj.firstChild.next might yield an object, where obj.children[0].next might not. This is intentional, since the children property doesn't use accNavigate to yield children and thus, when retrieving children using this method, we won't know whether accNavigate is implemented or not. Setting indexInParent anyway might lead to undesirable behaviour, so this is not something I'd like to do, unless we use accNavigate inside _get_children to check whether it's implemented or not.
2. This does not work for IAccessible objects that are created by events. For example, when navigating to the previous/next object from a focus object, this won't work when accNavigate isn't implemented. There is a simple workaround though, navigate to the parent object and back to the first child.

### Change log entry:
* Bug fixes
    + Fixed some cases where navigating to a next or previous IAccessible object didn't succeed even though a next/previous object was available. (#7178)

### CODE REVIEW

File: source/NVDAObjects/IAccessible/__init__.py [hunk 1]
```diff
@@ -531,17 +531,27 @@ def findOverlayClasses(self,clsList):
 			if clsList[0]==IAccessible and len(clsList)==3 and self.IAccessibleRole==oleacc.ROLE_SYSTEM_CLIENT and self.childCount==0:
 				clsList.insert(0,ContentGenericClient)
 
-	def __init__(self,windowHandle=None,IAccessibleObject=None,IAccessibleChildID=None,event_windowHandle=None,event_objectID=None,event_childID=None):
+	def __init__(self,windowHandle=None,IAccessibleObject=None,IAccessibleChildID=None,event_windowHandle=None,event_objectID=None,event_childID=None,parent=None,indexInParent=None):
+		"""
+		@param windowHandle: the window handle, if known.
```
Comment (dkager) [inline source/NVDAObjects/IAccessible/__init__.py]: This one and a few below it don't have a cap after the colon, while other params do.

File: source/NVDAObjects/IAccessible/__init__.py [hunk 2]
```diff
@@ -531,17 +531,27 @@ def findOverlayClasses(self,clsList):
 			if clsList[0]==IAccessible and len(clsList)==3 and self.IAccessibleRole==oleacc.ROLE_SYSTEM_CLIENT and self.childCount==0:
 				clsList.insert(0,ContentGenericClient)
 
-	def __init__(self,windowHandle=None,IAccessibleObject=None,IAccessibleChildID=None,event_windowHandle=None,event_objectID=None,event_childID=None):
+	def __init__(self,windowHandle=None,IAccessibleObject=None,IAccessibleChildID=None,event_windowHandle=None,event_objectID=None,event_childID=None,parent=None,indexInParent=None):
```
Comment (dkager) [inline source/NVDAObjects/IAccessible/__init__.py]: Ugh, this is a loooooooong line. Maybe break it up to 120 chars or so?

**Review by dkager — APPROVED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
Although I approved it from a coding point of view,  I would still like to see a clear practical use case in the wild before this is incubated or merged. I'm not sure if going only half way ends up actually being at all useful, except for  object navigation power users.

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [2 Feb 2018, 01:43 CET](https://github.com/nvaccess/nvda/pull/7647#issuecomment-362449971 "2018-02-02T00:43:39Z - Replied by Github Reply Comments"):
> Although I approved it from a coding point of view,  I would still like to see a clear practical use case in the wild before this is incubated or merged. I'm not sure if going only half way ends up actually being at all useful, except for  object navigation power users.

Giving a case in the wild is currently not very easy, except for if you forcefully disable UIA in NVDA and use NVDA"s MSAA fallback, for example in the settings app. Still, I agree that this working only in one direction is a bit puzzling.

I've also seen a case where accNavigate was implemented badly by Microsoft in a settings tab of the Active Directory management panel, but that case doesn't apply here, as accNavigate is implemented for most controls on that particular tab sheet, but not for all.

**Comment by leonardder** [comment]
After some deliberation, I've decided to close this for now:

1. As pointed out, there is currently no native MSAA case in the wild that allows testing this.
2. There might be a simpler method to avoid the underlying issue, for example when enforcing UIA on objects that fail with accNavigate. That also hasn't been tested yet.

I will leave the branch in tact for reference purposes.

**Pull request closed without merging**